enhancements/cluster-scope-secret-volumes/csi-driver-host-injections.md

```diff
@@ -0,0 +1,378 @@
+---
+title: share-secrets-and-configmaps-across-namespaces-via-a-csi-driver
+
+authors:
+  - "@gabemontero"
+
+reviewers:
+  - "@bparees"
+  - "@adambkaplan"
+  - "@derekwaynecarr"
+  - "@gnufied"
+  
+approvers:
+  - "@bparees"
+  - "@adambkaplan"
+  - "@derekwaynecarr"
+  - "@gnufied"
+  
+creation-date: 2020-03-17
+
+last-updated: 2020-03-30
+
+<!-- status: provisional|implementable|implemented|deferred|rejected|withdrawn|replaced -->
+status: provisional
+
+
+  
+---
+
+# Share Secrets And ConfigMaps Across Namespaces via a CSI Driver
+
+
+## Release Signoff Checklist
+
+- [/] Enhancement is `implementable`
+- [/] Design details are appropriately documented from clear requirements
+- [/] Test plan is defined
+- [/] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Open Questions [optional]
+
+> 1. Is a CSI `SecretProviderClass` based implementation recommended?  Received some varying guidance.  Also, per PR
+>review from @jsafrane, of which I can only concur based on my examination, the example at https://github.com/kubernetes-sigs/secrets-store-csi-driver
+>does not appear to handle refreshing of the data after initial provisioning.  So it does not appear to be something we
+>can use as is.  But do we want to take on adding such a feature upstream to that repo? **This question most
+>likely can be deferred until implementation.**
+> 2. What will be the precise means of interacting with the host filesystem.  The hostpath plugin actually uses the 
+>simple golang `os.MkdirAll` for example when creating mount access based volumes.  Can we do that in all our supported
+>scenarios?  We've also received suggestions around tmpfs or read only bind-mounts of a directory from the CSI driver.
+>**This question most likely can be deferred until implementation.**
+> 3. SELinux implications will need to be properly understood.  Minimally, a consuming Pod needs the correct SELinux 
+>context and fs permissions with bind-mounts for example.
+>**This question most likely can be deferred until implementation.**
+> 4. The EP document in https://github.com/openshift/enhancements/pull/169 needs to be reconciled with this document in
+>some form or fashion.
+
+## Summary
+
+This proposal will describe a [CSI driver](https://github.com/container-storage-interface/spec/blob/master/spec.md) based
+means for OpenShift cluster administrators to share `Secrets` / `ConfigMaps` defined in one namespace with
+other namesapces. 
+
+Any owners/authors/users of Pods in those namespaces can opt into the consumption of the content of those shared `Secrets` / 
```

Comment:
This is a little vague.  I don't think an average user should be able to say, "my namespace shares this cool secret/configmap".  We had namespace to namespace links in the first RBAC implementation and the referential integrity and discovery problems became too great to bear.  A single namespace makes sense and so does a list curated by cluster-admins, but arbitrary links seem bad.

Reply:
i think the mechanics of this are outside the scope of this EP (CSI driver) since the intent is that the CSI driver won't have any RBAC enforcement, it'll copy+mount whatever it is told to.

but when it does come to RBAC enforcement, i think our plan is that there's an admission controller which is configured by the admin, and that determines which namespaces/secrets/configmaps can be mounted by pods in which other namespaces.  So yeah, single/central configuration location, controlled by the admin.  Not random users offering up their secrets to random namespaces.

Reply:
see my slippery slope / new dividing line between this EP and #214 below

Reply:
> i think the mechanics of this are outside the scope of this EP (CSI driver) since the intent is that the CSI driver won't have any RBAC enforcement, it'll copy+mount whatever it is told to.

Yeah, I don't think that's a good premise. I think that a CSI driver needs to be aware of permission changes to prevent updates which should not be applied.

I think a goal of this should be a CSI driver that be generally programmed for use or building on top of.  If we aren't careful with API definitions now, we won't be able to do that without rebuilding what's here.

As I noted below, I don't think an API that is sympathetic to these needs is difficult.

Reply:
@gabemontero seems like this probably needs to be tightened up based on our discussion, it's now more like "users can opt into consuming content that a cluster-admin has chosen to make available"